gemviz/select_fields_tablemodel.py

```diff
@@ -0,0 +1,363 @@
+        """Column headers.  Called by QTableView."""
+        if role == QtCore.Qt.DisplayRole:
+            if orientation == QtCore.Qt.Horizontal:
+                return self.columnName(section)
+            else:
+                return str(section + 1)  # may want to alter at some point
+
+    def setData(self, index, value, role):
+        """Toggle the checkboxes.  Called by QTableView."""
+        if role == QtCore.Qt.CheckStateRole:
+            if index.column() in self.checkboxColumns:
+                self.setCheckbox(index, value)
+                return True
+        return False
+
+    def flags(self, index):
+        """Identify the checkbox cells.  Called by QTableView."""
+        original_flags = super().flags(index)
+        if index.column() in self.checkboxColumns:
+            # use a checkbox in this column
+            return original_flags | QtCore.Qt.ItemIsUserCheckable
+        return original_flags
+
+    # ------------ checkbox methods
+
+    def checkbox(self, index):
+        """Return the checkbox state."""
+        nm = self.columnName(index.column())  # selection name of THIS column
+        selection = self.selections.get(index.row())  # user selection
+        return QtCore.Qt.Checked if selection == nm else QtCore.Qt.Unchecked
+
+    def setCheckbox(self, index, state):
+        """Set the checkbox state."""
+        row, column = index.row(), index.column()
+        column_name = self.columnName(column)
+        checked = state == QtCore.Qt.Checked
+        prior = self.selections.get(row)
+        self.selections[row] = column_name if checked else None  # Rule 1
+        changes = self.selections[row] != prior
+        logger.debug("selections: %s", self.selections)
+
+        changes = self.applySelectionRules(index, changes)
+
+        if changes:
+            self.updateCheckboxes()
+
+        self.logCheckboxSelections()
+        logger.debug(self.plotFields())  # plotter should call plotFields()
+
+    def applySelectionRules(self, index, changes=False):
+        """Apply selection rules 2-4."""
+        row = index.row()
+        column_name = self.columnName(index.column())
+        for r, v in sorted(self.selections.items()):
+            if v is not None:
+                if self.columnNumber(v) in self.uniqueSelectionColumns:
+                    if r != row and column_name == v:
+                        self.selections[r] = None
+                        changes = True
+        return changes
+
+    def updateCheckboxes(self):
+        """Update checkboxes to agree with self.selections."""
+        if len(self.selections) > 0:
+            top, bottom = min(self.selections), max(self.selections)
+        else:
+            top, bottom = 0, self.rowCount() - 1
+        left, right = min(self.checkboxColumns), max(self.checkboxColumns)
+        logger.debug("corners: (%d,%d)  (%d,%d)", top, left, bottom, right)
+
+        # Re-evaluate the checkboxes bounded by the two corners (inclusive).
```

Comment:
That is the only part of the code that I did not understand, what are you doing with the corners?

Reply:
(I skipped it in mdaviz)